flink-libraries/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java

```diff
@@ -233,56 +235,100 @@ public ResultDescriptor executeQuery(SessionContext session, String query) throw
-				result.getTableSink(),
-				clusterClient);
-		} catch (Throwable t) {
-			// the result needs to be closed as long as
-			// it not stored in the result store
-			result.close();
-			throw t;
-		}
+			// create job graph with jars
+			final JobGraph jobGraph;
+			try {
+				jobGraph = createJobGraph(context, context.getSessionContext().getName() + ": " + query, table,
+						result.getTableSink(),
+						clusterClient);
+			} catch (Throwable t) {
+				// the result needs to be closed as long as
+				// it not stored in the result store
+				result.close();
+				throw t;
+			}
+
+			// store the result with a unique id (the job id for now)
+			final String resultId = jobGraph.getJobID().toString();
+			resultStore.storeDynamicResult(resultId, result);
+
+			// create execution
+			final Runnable program = () -> {
+				// we need to submit the job attached for now
+				// otherwise it is not possible to retrieve the reason why an execution failed
+				try {
+					clusterClient.run(jobGraph, context.getClassLoader());
+				} catch (ProgramInvocationException e) {
+					throw new SqlExecutionException("Could not execute table program.", e);
+				} finally {
+					try {
+						clusterClient.shutdown();
+					} catch (Exception e) {
+						// ignore
+					}
+				}
+			};
+
+			// start result retrieval
+			result.startRetrieval(program);
+
+			return new ResultDescriptor(resultId, resultSchema, result.isMaterialized());
 
-		// store the result with a unique id (the job id for now)
-		final String resultId = jobGraph.getJobID().toString();
-		resultStore.storeResult(resultId, result);
+		} else if (mergedEnv.getExecution().isBatchExecution()) {
 
-		// create execution
-		final Runnable program = () -> {
-			// we need to submit the job attached for now
-			// otherwise it is not possible to retrieve the reason why an execution failed
+			if (!mergedEnv.getExecution().isTableMode()) {
+				throw new SqlExecutionException("Batch queries can only work on table mode");
+			}
+
+			BatchResult batchResult = new BatchResult();
+			final JobGraph jobGraph = createJobGraph(
+					context,
+					context.getSessionContext().getName () + ": " + query,
+					table,
+					batchResult.getTableSink(),
+					clusterClient);
+
+			final String resultId = jobGraph.getJobID().toString();
+			final JobExecutionResult result;
 			try {
-				clusterClient.run(jobGraph, context.getClassLoader());
+				// Fetch the result synchronously.
+				result = clusterClient.run(jobGraph, context.getClassLoader());
 			} catch (ProgramInvocationException e) {
 				throw new SqlExecutionException("Could not execute table program.", e);
-			} finally {
+			}
+
+			ArrayList<byte[]> accResult = result.getAccumulatorResult(batchResult.getAccumulatorName());
+			if (accResult != null) {
 				try {
-					clusterClient.shutdown();
-				} catch (Exception e) {
-					// ignore
+					List<Row> resultTable = SerializedListAccumulator.deserializeList(accResult, batchResult.getSerializer());
```

Comment:
Can we move the deserialization logic into the result class? The executor should not need to handle internals of a result.